### PR TITLE
osmoutils: error tolerance helpers

### PR DESCRIPTION
**Pull request by pysel**
Closes: #6326 

## What is the purpose of the change

* Add tolerance comparison function between `osmomath.Dec`
* Add helper function to osmoassert to abstract assertion of equality in tests
* Migrate tests to use this new function

## Testing and Verifying

* Test for Dec comparison: https://github.com/osmosis-labs/osmosis/pull/6336/commits/d9f07a45ce956ab92f4816b6ad29e8e478451fde
* Migrated tests will check the helper

## Documentation and Release Note

N/A

### CODE REVIEW

File: osmomath/binary_search.go
Comment (pysel) [inline osmomath/binary_search.go:13]: note: I think this ordering is just way more comfortable

File: osmomath/binary_search.go
Comment (p0mvn) [inline osmomath/binary_search.go:148]: Would it be possible to somehow make a generic version shared by all of `Dec`, `BigDec`, `Int` etc?

**Review by p0mvn — COMMENTED** [review]
Nice! Some minor suggestions and requests.

Are you sure these are all the usage instances of `errTolerance` that we can refactor? It feels like there were more

**Comment by p0mvn** [comment]
> **Important Notice**
> 
> This PR modifies an in-repo Go module. It is one of:
> 
> * osmomath
> * osmoutils
> * x/ibc-hooks
> * x/epochs
> 
> The dependent Go modules, especially the root one, will have to be updated to reflect the changes. Failing to do so might cause e2e to fail.
> 
> Please follow the instructions below:
> 
> 1. Open https://github.com/osmosis/osmosis/actions/workflows/go-mod-auto-bump.yml
> 2. Provide the current branch name
> 3. On success, confirm if an automated commit corretly updated the go.mod and go.sum files
> 
> Please let us know if you need any help.

Just a thought: would be good to have an issue so that this doesn't comment if CI detects `go.mod` as updated

File: osmomath/binary_search.go
Comment (pysel) [reply osmomath/binary_search.go:148]: yes, I also thought about that one, totally possible to make it, should I make it in this PR?

Comment (p0mvn) [reply osmomath/binary_search.go:148]: Let's do in the next PR after merge please

File: osmomath/binary_search.go
Comment (mattverse) [inline osmomath/binary_search.go:147]: Can we add more test cases for all the logical branch in this method?
e.g) 
- it can either be round up or round down 
- it can either be additive or multiple tolerance.


And so on. I think we can make tdd depending on different combinations of paramters? Wdyt?

**Review by mattverse — COMMENTED** [review]
I like this idea! Asking for some test cases and we should be good to go

File: osmoutils/osmoassert/assertions.go
Comment (pysel) [reply osmoutils/osmoassert/assertions.go]: https://github.com/osmosis-labs/osmosis/pull/6336/commits/c928f7ca5538d36486f6f33a239da9c5d0059e85

Comment (pysel) [reply osmoutils/osmoassert/assertions.go]: https://github.com/osmosis-labs/osmosis/pull/6336/commits/4554ba0423f8eac5592a105e53f6e4862b715087

File: osmomath/binary_search.go
Comment (pysel) [reply osmomath/binary_search.go:174]: for context, next line is: `if e.AdditiveTolerance.IsZero() {`

if additive tolerance is zero (i.e., not acceptable), values should be equal to each other. The same comment exists in both `CompareBigDec` and `Compare[Int]`

File: osmomath/binary_search.go
Comment (pysel) [reply osmomath/binary_search.go:147]: > I think we can make tdd depending on different combinations of paramters? 

yes, there already existed a [test](https://github.com/osmosis-labs/osmosis/blob/ea8307152112afd5805c455057a50e94ed265b1d/osmomath/binary_search_test.go#L321-L372) for `Int` and `BigDec` comparison prior to this PR, which includes different configurations you mention. I added `Dec` comparison to that test in this commit: https://github.com/osmosis-labs/osmosis/pull/6336/commits/d9f07a45ce956ab92f4816b6ad29e8e478451fde

**Comment by pysel** [comment]
> Nice! Some minor suggestions and requests.
> 
> Are you sure these are all the usage instances of `errTolerance` that we can refactor? It feels like there were more

hmm, you are right. I found more instances, but they directly use `require` package to assert comparison([example](https://github.com/osmosis-labs/osmosis/blob/ea8307152112afd5805c455057a50e94ed265b1d/osmomath/binary_search_test.go#L219)).
I refactored Equal() function to take `testing.T` and not `testSuite` here: https://github.com/osmosis-labs/osmosis/pull/6336/commits/80b3ad8c87c2dfa5ca15d6af6e40873c7e9c0c9b.

thanks for noting that out

cc: @p0mvn @mattverse

**Comment by pysel** [comment]
question to reviewers: why do we need this comparison - https://github.com/osmosis-labs/osmosis/blob/main/x/twap/api_test.go#L899-L904? the result is not checked. if it should not be checked, we can remove those lines, if the result should be 0, it is unchecked. 

I tried to check the result value: https://github.com/osmosis-labs/osmosis/pull/6336/files#diff-f86385179fdd51029d8446f3063a7a6e71b8ac928c97e794ccad9b96f82ee28cR900, test fails

cc: @mattverse @p0mvn

**Comment by p0mvn** [comment]
> question to reviewers: why do we need this comparison - https://github.com/osmosis-labs/osmosis/blob/main/x/twap/api_test.go#L899-L904? the result is not checked. if it should not be checked, we can remove those lines, if the result should be 0, it is unchecked.
> 
> I tried to check the result value: https://github.com/osmosis-labs/osmosis/pull/6336/files#diff-f86385179fdd51029d8446f3063a7a6e71b8ac928c97e794ccad9b96f82ee28cR900, test fails
> 
> cc: @mattverse @p0mvn

Please make an issue documenting all the findings and don't worry about it in this PR as it is out of scope

**Review by p0mvn — APPROVED** [review]
LGTM.

Could you please merge main into this PR to make sure nothing breaks? I think we can merge this after

**Comment by pysel** [comment]
hey @p0mvn! issue created: https://github.com/osmosis-labs/osmosis/issues/6348, main merged

**Comment by p0mvn** [comment]
> hey @p0mvn! issue created: #6348, main merged

Please review CI and re-ping once ready

**Review by p0mvn — APPROVED** [review]

**Pull request merged**
